### PR TITLE
Closes 5719: Forced mobile cache when WpDiscuz is activated.

### LINKED ISSUE #5719: wpDiscuz needs a separate cache for mobile
**Before submitting an issue please check that you’ve completed the following steps:**
- [x] Made sure you’re on the latest version `3.12.5.2`
- [x] Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
wpDiscuz uses `wp_is_mobile()` to control adding some assets to the front end. 

Here is an example related the Rick Editor settings:

image

File: `/class.WpdiscuzCore.php:1235`.

Extend of the customizations done depending on the setting:

```
$loadQuill = $this->options->form["richEditor"] === "both" || (!wp_is_mobile() && $this->options->form["richEditor"] === "desktop");
if (!$loadQuill) {
    $combo_js = "-no_quill";
    $combo_css = "-no_quill";
}
if ($loadQuill) {
    wp_add_inline_script("wpdiscuz-combo-js", $this -> options -> editorOptions(), "before");
}
if ($loadQuill) {
    wp_register_style("quill-css", plugins_url(WPDISCUZ_DIR_NAME. "/assets/third-party/quill/quill.snow$suf.css"), null, "1.3.6");
    wp_enqueue_style("quill-css");
    wp_register_script("quill", plugins_url("/assets/third-party/quill/quill$suf.js", __FILE__), ["wpdiscuz-cookie-js"], "1.3.6", true);
    wp_enqueue_script("quill");
    wp_add_inline_script("quill", $this -> options -> editorOptions(), "before");
    wp_register_script("wpd-editor", plugins_url("/assets/js/wpd-editor$suf.js", __FILE__), ["quill"], "1.3.6", true);
    wp_enqueue_script("wpd-editor");
}
if (!$loadQuill && $this -> options -> form["enableQuickTags"]) {
    wp_enqueue_script("quicktags");
    wp_register_script("wpdiscuz-quicktags", plugins_url("/assets/third-party/quicktags/wpdiscuz-quictags.js", __FILE__), [$this -> options -> general["loadComboVersion"] ? "wpdiscuz-combo-js" : "wpdiscuz-ajax-js"], $this -> version, true);
    wp_enqueue_script("wpdiscuz-quicktags");
}
```

There are other occurrences controlling the output depending on if `wp_is_mobile()` is set or not. 

This could break features on the page and render the wrong style.

Plugin URI - `wpdiscuz/class.WpdiscuzCore.php`.

I will leave it to QA to see if we should target locating specific features through a compatibility file or just add `wpDiscuz` as part of the [get_mobile_plugins()](https://github.com/wp-media/wp-rocket/blob/1246d406f0523799a5d084a94e61fabf33b3405f/inc/classes/subscriber/third-party/plugins/class-mobile-subscriber.php#L331-L353) list.

**To Reproduce**
1. Enable wpDiscuz
2. Enable the above-mentioned setting as an example
3. Compare the output of a page desktop vs. mobile to see that different assets are loaded

**Expected behavior**
We should accommodate scenarios where wpDiscuz will output additional or different assets on the frontend. 

**Screenshots**
N/A 

**Additional context**
Ticket - https://secure.helpscout.net/conversation/2132824521/396631?folderId=3864735

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by CrochetFeve0251**
## Description

Fix an issue concerning WpDiscuz that loads different css on mobile and desktop.

For that  we forced mobile cache when the plugin is active.

Fixes #5719

## Type of change

- [ ] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

No.

## How Has This Been Tested?

- [ ] Automated Test

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: inc/ThirdParty/Plugins/WpDiscuz.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	 * Subscriber for wpDiscuz.
+	 *
+	 * @return array
+	 */
+	public static function get_subscribed_events() {
+		if ( ! defined( 'WPDISCUZ_DIR_NAME' ) ) {
+			return [];
+		}
+
+		return [ 'option_' . rocket_get_constant( 'WP_ROCKET_SLUG' ) => 'force_mobile_cache' ];
```
Comment (jeawhanlee) [inline inc/ThirdParty/Plugins/WpDiscuz.php]: I think we can use the `pre_get_rocket_option_{$option_key}` filter here which in this case would be `pre_get_rocket_option_do_caching_mobile_files`.
Also shouldn't we force mobile cache - `cache_mobile` as `do_caching_mobile_files` is dependent on `cache_mobile`.

We can use the `WP_Rocket\ThirdParty\ReturnTypesTrait` to return bool

**Review by jeawhanlee — COMMENTED** [review]
Tests are failing also.

**Review by jeawhanlee — APPROVED** [review]

**Comment by piotrbak** [comment]
@CrochetFeve0251 @jeawhanlee We have a function that should cover this scenario:
https://github.com/wp-media/wp-rocket/blob/1246d406f0523799a5d084a94e61fabf33b3405f/inc/classes/subscriber/third-party/plugins/class-mobile-subscriber.php#L331

I think it's better to use this one instead of creating new comaptibility. Unless I'm missing something, what do you think?

CC @Mai-Saad

**Comment by piotrbak** [comment]
@Tabrisrp What do you think about this? Should we simply add this plugin to the old architecture code?

If not, I'd probably prefer to block this PR as Separate Cache for Mobiles will be a default setting soon, so there's no need of rewriting this part.

**Review by Mai-Saad — APPROVED** [review]
Working as expected notice that we agreed that following scenario is fine for now (same on trunk and 3.11.5):
1- compatibility plugin is active
2- fresh install and activation of wpr => separate mobile cache enabled by default and dimmed
3- deactivate compatibility plugin => separate mobile cache is still enabled but not dimm
[testrail-report-461.pdf](https://github.com/wp-media/wp-rocket/files/11029494/testrail-report-461.pdf)

**Pull request merged**